lens/interface.go

```diff
@@ -78,21 +85,56 @@ type MessageExecution struct {
 	Implicit bool
 }
 
-type ExecutedMessage struct {
-	Cid           cid.Cid
-	Height        abi.ChainEpoch
-	Message       *types.Message
-	Receipt       *types.MessageReceipt
-	BlockHeader   *types.BlockHeader
-	Blocks        []cid.Cid // blocks this message appeared in
-	Index         uint64    // Message and receipt sequence in tipset
-	FromActorCode cid.Cid   // code of the actor the message is from
-	ToActorCode   cid.Cid   // code of the actor the message is to
-	GasOutputs    vm.GasOutputs
-}
-
 type BlockMessages struct {
 	Block        *types.BlockHeader     // block messages appeared in
 	BlsMessages  []*types.Message       // BLS messages in block `Block`
 	SecpMessages []*types.SignedMessage // SECP messages in block `Block`
 }
+
+type BlockMessageReceipts struct {
+	Block                 *types.BlockHeader
+	Message               []types.ChainMsg
+	Receipt               []*types.MessageReceipt
+	MessageExecutionIndex map[types.ChainMsg]int
+}
+
+func (bmr *BlockMessageReceipts) Iterator() (*MessageReceiptIterator, error) {
+	if len(bmr.Message) != len(bmr.Receipt) {
+		return nil, fmt.Errorf("invalid construction, expected equal number receipts (%d) and messages (%d)", len(bmr.Receipt), len(bmr.Message))
+	}
+	return &MessageReceiptIterator{
+		idx:      0,
+		msgs:     bmr.Message,
+		receipts: bmr.Receipt,
+		exeIdx:   bmr.MessageExecutionIndex,
+	}, nil
+}
+
+type MessageReceiptIterator struct {
+	idx      int
+	msgs     []types.ChainMsg
+	receipts []*types.MessageReceipt
+	exeIdx   map[types.ChainMsg]int
+}
+
+func (mri *MessageReceiptIterator) HasNext() bool {
+	if mri.idx < len(mri.msgs) {
+		return true
+	}
+	return false
+}
+
+func (mri *MessageReceiptIterator) Next() (types.ChainMsg, int, *types.MessageReceipt) {
```

Comment:
```suggestion
func (mri *MessageReceiptIterator) Next() (types.ChainMsg, *types.MessageReceipt) {
```
is the index necessary as part of the return value? i did a quick scan of where this is being called and only saw it once [here](https://github.com/filecoin-project/lily/pull/1040/files#diff-9134a88da118334f295ecafd8ba9e83fb372a7922b753bb32f7f4b3099cf2c1dR72).

could the index also be accessed as:
```go
msg, rec := itr.Next()
idx := itr.exeIdx[msg]
```